# Part 2: Data Manipulation

In this tutorial, we're going to be using the [Starbucks menu nutrional facts](https://www.kaggle.com/starbucks/starbucks-menu/data) dataset which consists of two files - one for beverages and one for food. This dataset is located in the `data/` folder of this repository. You can download the original dataset via [Kaggle](https://www.kaggle.com/), which is an online community of data scientists and machine learning practitioners. Within Kaggle, you can find a wide variety of open-source datasets and public machine learning competitions. 

A fill-in-the-blank version of this notebook can be accessed on [Google Colab](https://colab.research.google.com/drive/1-de8UlYtOMilR8pFvSJvn4uApXOs4mzC).

### Importing Dependencies

We'll be using the following packages:

- [pandas](https://pandas.pydata.org/) - good for data manipulation and analysis
- [numpy](https://numpy.org/) - good for scientific computing 
- [matplotlib](https://matplotlib.org/) - plotting package
- [seaborn](https://seaborn.pydata.org/) - plotting package (an extension of matplotlib)

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

pd.options.mode.chained_assignment = None

/usr/local/lib/python3.7/site-packages/matplotlib/__init__.py:886: MatplotlibDeprecationWarning: 
examples.directory is deprecated; in the future, examples will be found relative to the 'datapath' directory.
  "found relative to the 'datapath' directory.".format(key))


### Loading the Data

Our Starbucks dataset consists of two files: 

1. nutritional facts about Starbucks beverages
2. nutritional facts about Starbucks food items 

We'll be loading these files in as two separate dataframes. 

### What is a dataframe?

<img src="assets/imgs/dataframe_anatomy.png" width="50%" align="right">

A Pandas DataFrame is a two-dimensional data structure. It consists of three components: 

1. rows: represents separate entries/samples of the dataset
2. columns: represents features
3. data: populates the cells of the dataframe


You can create a dataframe from scratch:

```
pd.DataFrame({
    'drink': ['coffee', 'tea', 'latte'], 
    'price': [1.50, 2, 3.50]
})
```

You can also create a dataframe by loading in a csv file, excel spreadsheet, or table from a SQL database: 

```
pd.read_csv('dataset.csv')
```

```
pd.read_excel('dataset.xls')
```

Let's start by loading our Starbucks drink menu. It's a csv file so we'll be using `read_csv()`.

In [2]:
starbucks_drink = pd.read_csv("data/starbucks_drink_menu.csv")

Before we start analyzing our dataset, let's first take a look at how many rows and columns we're dealing with. You can get this information by applying `.shape` to your dataframe. This will return a tuple of (`n_rows`, `n_cols`).

In [3]:
starbucks_drink.shape

(240, 19)

Let's also take a peek at the first 5 rows of our dataset. We do this using the `.head()` method.

In [4]:
starbucks_drink.head()

beverage        beverage_category drink_size milk_type  calories  \
0    Brewed Coffee                   Coffee      Short       NaN         3   
1    Brewed Coffee                   Coffee       Tall       NaN         4   
2    Brewed Coffee                   Coffee     Grande       NaN         5   
3    Brewed Coffee                   Coffee      Venti       NaN         5   
4  Caffè Americano  Classic Espresso Drinks      Short       NaN         5   

  caffeine_in_mg  carbs_in_g  sugars_in_g  protein_in_g  fat_in_g  fibre_in_g  \
0            175           5            0           0.3       0.1           0   
1            260          10            0           0.5       0.1           0   
2            330          10            0           1.0       0.1           0   
3            410          10            0           1.0       0.1           0   
4             75           5            0           0.4       0.0           0   

   cholesterol_in_mg  saturated_fat_in_g  sodium_in_mg  trans_fat_in_g  \
0                  0                 0.0             0             0.0   
1                  0                 0.0             0             0.0   
2                  0                 0.0             0             0.0   
3                  0                 0.0             0             0.0   
4                  1                 0.0             0             0.0   

  vit_a_dv vit_c_dv iron_dv calcium_dv  
0       0%       0%      0%         0%  
1       0%       0%      0%         0%  
2       0%       0%      0%         0%  
3       0%       0%      0%         2%  
4       0%       0%      0%         0%

If we wanted to randomly sample rows, we can use a method called `.sample()`. You can decide how many rows you want to sample by adjusting `n`.

In [5]:
starbucks_drink.sample(n=3)

beverage      beverage_category  \
124  Iced Brewed Coffee (With Milk & Classic Syrup)  Shaken Iced Beverages   
226                            Tazo® Chai Tea Latte       Tazo® Tea Drinks   
2                                     Brewed Coffee                 Coffee   

    drink_size milk_type  calories caffeine_in_mg  carbs_in_g  sugars_in_g  \
124      short   2% Milk        90            NaN          25           18   
226     grande   Soymilk       220             95          90           37   
2       Grande       NaN         5            330          10            0   

     protein_in_g  fat_in_g  fibre_in_g  cholesterol_in_mg  \
124           2.0       1.0           0                 18   
226           6.0       3.5           1                 41   
2             1.0       0.1           0                  0   

     saturated_fat_in_g  sodium_in_mg  trans_fat_in_g vit_a_dv vit_c_dv  \
124                 0.0             5             0.5       2%       0%   
226                 0.0             0             0.4       8%       0%   
2                   0.0             0             0.0       0%       0%   

    iron_dv calcium_dv  
124   0.00%         6%  
226     10%        25%  
2        0%         0%

### Which columns are in our dataset?

We can see which columns are in our dataset, and look at their datatypes using `dtypes`. 

In [6]:
starbucks_drink.dtypes

beverage               object
beverage_category      object
drink_size             object
milk_type              object
calories                int64
caffeine_in_mg         object
carbs_in_g              int64
sugars_in_g             int64
protein_in_g          float64
fat_in_g              float64
fibre_in_g              int64
cholesterol_in_mg       int64
saturated_fat_in_g    float64
sodium_in_mg            int64
trans_fat_in_g        float64
vit_a_dv               object
vit_c_dv               object
iron_dv                object
calcium_dv             object
dtype: object

We can see that our dataset has a combination of `int64`, `float64` and `object` datatypes. The `object` datatype  suggests that the column is either a string type (categorical) or mixed (e.g., mixture of floats and NaNs). A NaN is a "none" type, which means that the value is missing.   

### Selecting a Column From a Pandas DataFrame

Let's get a single column from our dataframe. We can do this by identifying the column name in square brackets:

```
data['col_name']
```
The output is a [Pandas Series](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html) which comes with its own special functions. 

In [7]:
sample_data = starbucks_drink.head()
sample_data['calories']

0    3
1    4
2    5
3    5
4    5
Name: calories, dtype: int64

Some useful Pandas Series methods include: `.mean()`, `.max()`, `.min()`, `.describe()`, `.value_counts()`. Try these out in the cell below. 

In [8]:
sample_data['calories'].mean()

4.4

In [9]:
sample_data['calories'].describe()

count    5.000000
mean     4.400000
std      0.894427
min      3.000000
25%      4.000000
50%      5.000000
75%      5.000000
max      5.000000
Name: calories, dtype: float64

### Selecting a Row From a Pandas DataFrame

What if we wanted to look at one row in our dataframe? Rows can be accessed by their index. A dataframe's index typically starts at 0 and increments with each row. We can access a dataframe's row using `.iloc[index_number]` where index_number represents the index of the row of interest.

Let's get the first row of our sample dataframe. In this case, the index will be 0. 

In [10]:
sample_data.iloc[0]

beverage              Brewed Coffee
beverage_category            Coffee
drink_size                    Short
milk_type                       NaN
calories                          3
caffeine_in_mg                  175
carbs_in_g                        5
sugars_in_g                       0
protein_in_g                    0.3
fat_in_g                        0.1
fibre_in_g                        0
cholesterol_in_mg                 0
saturated_fat_in_g                0
sodium_in_mg                      0
trans_fat_in_g                    0
vit_a_dv                         0%
vit_c_dv                         0%
iron_dv                          0%
calcium_dv                       0%
Name: 0, dtype: object

With `.iloc[]`, you can also access specific elements in the dataframe. You can do this by passing in a list of length 2. The first element represents the row position, while the second element represents the column position.

In [11]:
sample_data.iloc[0,0]

'Brewed Coffee'

Let's get `drink_size`. Remember, index positions starts at 0. The column `drink_size` is the 3rd column in our dataframe, so its position will be 2.

In [12]:
sample_data.iloc[0,2]

'Short'

### Adding Columns to a Pandas DataFrame

We can create a new column in our dataframe by declaring a new list as a column. The length of the list must be equal to the number of rows in the dataframe. In our `sample_data`, we have 5 rows so our list must be of length 5. Let's take a look at the example below, where we create a new column indicating whether a Starbucks beverage is in stock.

In [13]:
in_stock = [True, False, True, False, True]

sample_data['in_stock'] = in_stock

What if all rows in our new column have the same value? We can do this simply declaring a single value as a column.

In [14]:
sample_data['available_in_canada'] = True 

Let's take a look at our new columns. Scroll right.

In [15]:
sample_data

beverage        beverage_category drink_size milk_type  calories  \
0    Brewed Coffee                   Coffee      Short       NaN         3   
1    Brewed Coffee                   Coffee       Tall       NaN         4   
2    Brewed Coffee                   Coffee     Grande       NaN         5   
3    Brewed Coffee                   Coffee      Venti       NaN         5   
4  Caffè Americano  Classic Espresso Drinks      Short       NaN         5   

  caffeine_in_mg  carbs_in_g  sugars_in_g  protein_in_g  fat_in_g  \
0            175           5            0           0.3       0.1   
1            260          10            0           0.5       0.1   
2            330          10            0           1.0       0.1   
3            410          10            0           1.0       0.1   
4             75           5            0           0.4       0.0   

          ...           cholesterol_in_mg  saturated_fat_in_g  sodium_in_mg  \
0         ...                           0                 0.0             0   
1         ...                           0                 0.0             0   
2         ...                           0                 0.0             0   
3         ...                           0                 0.0             0   
4         ...                           1                 0.0             0   

   trans_fat_in_g  vit_a_dv vit_c_dv iron_dv calcium_dv in_stock  \
0             0.0        0%       0%      0%         0%     True   
1             0.0        0%       0%      0%         0%    False   
2             0.0        0%       0%      0%         0%     True   
3             0.0        0%       0%      0%         2%    False   
4             0.0        0%       0%      0%         0%     True   

   available_in_canada  
0                 True  
1                 True  
2                 True  
3                 True  
4                 True  

[5 rows x 21 columns]

### Removing Columns from a Pandas DataFrame

To remove a column from our dataframe, we can use the [drop()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop.html) method. Inside `drop()`, we can specify either a single column or a list of columns that we want to remove. Let's try removing `beverage`, `beverage_category`, and `drink_size`. 

In [16]:
sample_data.drop(columns=['beverage', 'beverage_category', 'drink_size'])

milk_type  calories caffeine_in_mg  carbs_in_g  sugars_in_g  protein_in_g  \
0       NaN         3            175           5            0           0.3   
1       NaN         4            260          10            0           0.5   
2       NaN         5            330          10            0           1.0   
3       NaN         5            410          10            0           1.0   
4       NaN         5             75           5            0           0.4   

   fat_in_g  fibre_in_g  cholesterol_in_mg  saturated_fat_in_g  sodium_in_mg  \
0       0.1           0                  0                 0.0             0   
1       0.1           0                  0                 0.0             0   
2       0.1           0                  0                 0.0             0   
3       0.1           0                  0                 0.0             0   
4       0.0           0                  1                 0.0             0   

   trans_fat_in_g vit_a_dv vit_c_dv iron_dv calcium_dv  in_stock  \
0             0.0       0%       0%      0%         0%      True   
1             0.0       0%       0%      0%         0%     False   
2             0.0       0%       0%      0%         0%      True   
3             0.0       0%       0%      0%         2%     False   
4             0.0       0%       0%      0%         0%      True   

   available_in_canada  
0                 True  
1                 True  
2                 True  
3                 True  
4                 True

To save this trimmed down dataframe, we would need to assign it a variable name. Alternatively, we can set the parameter `inplace=True` inside `drop()` which will permanently alter the columns of our `sample_data` dataframe.

### Removing Duplicates from a Pandas DataFrame

Let's take a look at the `beverage` column. We can see that there are quite a few duplicates of "Brewed Coffee". Let's say we wanted to have just one entry for this beverage type. We can drop all duplicates using the [drop_duplicates()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop_duplicates.html) method. 

If we apply drop_duplicates() without specifying any parameters, it will return the same dataframe. 

In [17]:
sample_data.drop_duplicates()

beverage        beverage_category drink_size milk_type  calories  \
0    Brewed Coffee                   Coffee      Short       NaN         3   
1    Brewed Coffee                   Coffee       Tall       NaN         4   
2    Brewed Coffee                   Coffee     Grande       NaN         5   
3    Brewed Coffee                   Coffee      Venti       NaN         5   
4  Caffè Americano  Classic Espresso Drinks      Short       NaN         5   

  caffeine_in_mg  carbs_in_g  sugars_in_g  protein_in_g  fat_in_g  \
0            175           5            0           0.3       0.1   
1            260          10            0           0.5       0.1   
2            330          10            0           1.0       0.1   
3            410          10            0           1.0       0.1   
4             75           5            0           0.4       0.0   

          ...           cholesterol_in_mg  saturated_fat_in_g  sodium_in_mg  \
0         ...                           0                 0.0             0   
1         ...                           0                 0.0             0   
2         ...                           0                 0.0             0   
3         ...                           0                 0.0             0   
4         ...                           1                 0.0             0   

   trans_fat_in_g  vit_a_dv vit_c_dv iron_dv calcium_dv in_stock  \
0             0.0        0%       0%      0%         0%     True   
1             0.0        0%       0%      0%         0%    False   
2             0.0        0%       0%      0%         0%     True   
3             0.0        0%       0%      0%         2%    False   
4             0.0        0%       0%      0%         0%     True   

   available_in_canada  
0                 True  
1                 True  
2                 True  
3                 True  
4                 True  

[5 rows x 21 columns]

This is because there are no two rows that are exactly identical across all columns. However, if we specify the subset of columns, it will drop the rows that contain duplicates within that column or list of columns. When we define the subset to be [`beverage`,`beverage_category`], it will remove all duplicate rows containing the same combination of `beverage` and `beverage_category`. You can decide which rows to drop using the `keep` parameter. By default, `keep` is set to 'first' which means that the first row is considered to be unique while the rest of the same value are duplicates. 

In the example below, we keep the first row containing "Brewed Coffee" and "Coffee", while all subsequent rows get dropped. 

In [18]:
sample_data.drop_duplicates(subset=['beverage', 'beverage_category'])

beverage        beverage_category drink_size milk_type  calories  \
0    Brewed Coffee                   Coffee      Short       NaN         3   
4  Caffè Americano  Classic Espresso Drinks      Short       NaN         5   

  caffeine_in_mg  carbs_in_g  sugars_in_g  protein_in_g  fat_in_g  \
0            175           5            0           0.3       0.1   
4             75           5            0           0.4       0.0   

          ...           cholesterol_in_mg  saturated_fat_in_g  sodium_in_mg  \
0         ...                           0                 0.0             0   
4         ...                           1                 0.0             0   

   trans_fat_in_g  vit_a_dv vit_c_dv iron_dv calcium_dv in_stock  \
0             0.0        0%       0%      0%         0%     True   
4             0.0        0%       0%      0%         0%     True   

   available_in_canada  
0                 True  
4                 True  

[2 rows x 21 columns]

In the example below, we set `keep` to be 'last' so we keep the last row containing "Brewed Coffee" and "Coffee" while all preceding rows get dropped.

In [19]:
sample_data.drop_duplicates(subset=['beverage', 'beverage_category'], keep="last")

beverage        beverage_category drink_size milk_type  calories  \
3    Brewed Coffee                   Coffee      Venti       NaN         5   
4  Caffè Americano  Classic Espresso Drinks      Short       NaN         5   

  caffeine_in_mg  carbs_in_g  sugars_in_g  protein_in_g  fat_in_g  \
3            410          10            0           1.0       0.1   
4             75           5            0           0.4       0.0   

          ...           cholesterol_in_mg  saturated_fat_in_g  sodium_in_mg  \
3         ...                           0                 0.0             0   
4         ...                           1                 0.0             0   

   trans_fat_in_g  vit_a_dv vit_c_dv iron_dv calcium_dv in_stock  \
3             0.0        0%       0%      0%         2%    False   
4             0.0        0%       0%      0%         0%     True   

   available_in_canada  
3                 True  
4                 True  

[2 rows x 21 columns]

### Subsetting a Dataframe

We can use [boolean indexing](https://www.dunderdata.com/blog/selecting-subsets-of-data-in-pandas-part-2) to filter entries of a dataframe based on conditions from a set of columns. Let's say we want to see which beverages have more than 450 calories. We can create a boolean condition:

```
starbucks_drink['calories'] > 450
```

which will return True for beverages that contain more than 450 calories and False for those that contain 450 calories or less. If we pass this condition inside the dataframe, it will return only the rows that meet this condition (i.e., return True). Let's try it out:

In [20]:
starbucks_drink[starbucks_drink['calories'] > 450]

beverage  \
138  White Chocolate Mocha (Without Whipped Cream)   
166              Java Chip (Without Whipped Cream)   
220  White Chocolate Mocha (Without Whipped Cream)   

               beverage_category drink_size   milk_type  calories  \
138    Signature Espresso Drinks      venti     2% Milk       510   
166  Frappuccino® Blended Coffee     grande  Whole Milk       460   
220    Signature Espresso Drinks      venti     Soymilk       460   

    caffeine_in_mg  carbs_in_g  sugars_in_g  protein_in_g  fat_in_g  \
138            150         330           74          19.0      15.0   
166            145         340           84           7.0      10.0   
220            150         290           64          16.0      13.0   

     fibre_in_g  cholesterol_in_mg  saturated_fat_in_g  sodium_in_mg  \
138           0                 77                 0.2            35   
166           2                 90                 0.2            15   
220           1                 70                 0.0             5   

     trans_fat_in_g vit_a_dv vit_c_dv iron_dv calcium_dv  
138             9.0      20%       4%      2%        60%  
166             7.0       6%       0%     35%        15%  
220             6.0      15%       2%     20%        60%

Now, let's say we want to get all beverages that contain more than 450 calories and at least 1g of fiber. Our boolean condition will be:

```
(starbucks_drink['calories'] > 450) & (starbucks_drink['fiber_in_g'] >= 1)
```

Let's see what the results are:

In [21]:
starbucks_drink[(starbucks_drink['calories'] > 450)&(starbucks_drink['fibre_in_g'] >= 1)]

beverage  \
166              Java Chip (Without Whipped Cream)   
220  White Chocolate Mocha (Without Whipped Cream)   

               beverage_category drink_size   milk_type  calories  \
166  Frappuccino® Blended Coffee     grande  Whole Milk       460   
220    Signature Espresso Drinks      venti     Soymilk       460   

    caffeine_in_mg  carbs_in_g  sugars_in_g  protein_in_g  fat_in_g  \
166            145         340           84           7.0      10.0   
220            150         290           64          16.0      13.0   

     fibre_in_g  cholesterol_in_mg  saturated_fat_in_g  sodium_in_mg  \
166           2                 90                 0.2            15   
220           1                 70                 0.0             5   

     trans_fat_in_g vit_a_dv vit_c_dv iron_dv calcium_dv  
166             7.0       6%       0%     35%        15%  
220             6.0      15%       2%     20%        60%

Great! The resulting dataframe gives us beverages that have more than 450 calories and at least 1g of fiber. 

### Cleaning a Column

Some of our columns have mixed datatypes. The column `caffeine_in_mg` has mainly numerical values, but there are a few values that are "Varies"/"varies" which is of type string. We won't be able to do any numerical computations on this column until we make sure all values are either a float or integer. Let's convert these string values to `None`. We'll first convert the column to lowercase (so that "Varies" becomes "varies") then replace "varies" with None. Check out how it works: 

In [22]:
starbucks_drink['caffeine_in_mg'] = starbucks_drink['caffeine_in_mg'].str.lower().replace('varies', None)

We'll convert this column to type "float". We can do this using `astype(float)`. If we wanted to convert it to a string, we would apply `astype(str)`.

In [23]:
starbucks_drink['caffeine_in_mg'] = starbucks_drink['caffeine_in_mg'].astype(float)

Now that our column `caffeine_in_mg` is of type float, we can calculate the mean and max.

In [24]:
mean_caffeine = starbucks_drink['caffeine_in_mg'].mean()
max_caffeine = starbucks_drink['caffeine_in_mg'].max()

print(f"Mean amount of caffeine in mg: {mean_caffeine}")
print(f"Max caffeine in mg: {max_caffeine}")

Mean amount of caffeine in mg: 87.00836820083683
Max caffeine in mg: 410.0


The mean is a long floating point number. Let's round it down to 2 decimals.

In [25]:
print(f"Mean amount of caffeine in mg: {round(mean_caffeine, 2)}")

Mean amount of caffeine in mg: 87.01


### Applying an Operation to a Column

If you take a look at the unique values in the `drinnk_size` column, you'll notice that there are duplicates because there are 2 spelling variations of a given drink size - one is lower case while the other is capitalized.

In [26]:
starbucks_drink['drink_size'].unique()

array(['Short', 'Tall', 'Grande', 'Venti', 'short', 'tall', 'grande',
       'venti'], dtype=object)

To keep things consistent, let's make all drink size names lower case. We'll start by creating a function that passes in some text and makes it lower case.

In [27]:
def lower_case_text(text):
    return text.lower()

drink_size = 'Tall'
lower_case_text(drink_size)

'tall'

To apply our function `lower_case_text()` to the `drink_size` column, we can use the `apply()` method which applies the operation to all rows of the column. 

In [28]:
starbucks_drink['drink_size'] = starbucks_drink['drink_size'].apply(lower_case_text)

In [29]:
starbucks_drink['drink_size'].unique()

array(['short', 'tall', 'grande', 'venti'], dtype=object)

Awesome! Our drink size names have now been standardized. Let's take a look at the frequency counts of each drink size using `value_counts()`. 

In [30]:
starbucks_drink['drink_size'].value_counts()

grande    67
tall      64
short     58
venti     51
Name: drink_size, dtype: int64

We can see that "grande" is the most common size that Starbucks offers, while Venti is the least common.

### Renaming a Column 

We can access the list of our dataframe's column names using its "columns" attribute. 

In [31]:
starbucks_drink.columns

Index(['beverage', 'beverage_category', 'drink_size', 'milk_type', 'calories',
       'caffeine_in_mg', 'carbs_in_g', 'sugars_in_g', 'protein_in_g',
       'fat_in_g', 'fibre_in_g', 'cholesterol_in_mg', 'saturated_fat_in_g',
       'sodium_in_mg', 'trans_fat_in_g', 'vit_a_dv', 'vit_c_dv', 'iron_dv',
       'calcium_dv'],
      dtype='object')

Let's say we want to change one of our columns from `fibre_in_g` (the British spelling) to `fiber_in_g` (the American spelling). There are two approaches to making this change. The first approach is to use the [rename](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rename.html) method, passing in a dictionary with the key as the current column name and the value as the new column name. 

In [32]:
starbucks_drink = starbucks_drink.rename(columns={'fibre_in_g': 'fiber_in_g'})

The second approach is to reassign `starbucks.columns` to a new list of column names. This is useful if we want to change multiple column names. 

In [33]:
sample_data = starbucks_drink.sample(n=3)

sample_data.columns = ['bev', 'bev_cat', 'drink_size', 'milk', 'cals',
       'caffeine', 'carbs', 'sugars', 'protein', 'fat', 'fiber', 'chol', 'sat_fat',
       'sodium_in_mg', 'trans_fat_in_g', 'vit_a_dv', 'vit_c_dv', 'iron_dv', 'calcium_dv']

sample_data

bev                     bev_cat  \
104  Strawberries & Crème (Without Whipped Cream)  Frappuccino® Blended Crème   
187      Vanilla Latte (Or Other Flavoured Latte)     Classic Espresso Drinks   
173          Vanilla Bean (Without Whipped Cream)  Frappuccino® Blended Crème   

    drink_size         milk  cals  caffeine  carbs  sugars  protein  fat  \
104     grande  Nonfat Milk   230       0.0    190      52      4.0  0.2   
187       tall      Soymilk   160      75.0     95      20      7.0  4.0   
173      short   Whole Milk   200       0.0    160      38      3.0  3.5   

     fiber  chol  sat_fat  sodium_in_mg  trans_fat_in_g vit_a_dv vit_c_dv  \
104      0    53      0.0             0             0.1       8%       6%   
187      1    23      0.0             0             0.5      10%       0%   
173      0    39      0.1            10             2.0       6%       0%   

    iron_dv calcium_dv  
104      4%        15%  
187     15%        30%  
173      0%        10%

### Appending Two DataFrames Together

Let's say we have several dataframes with the same column format that we want to combine into a single dataframe. We can do this using a technique called [concatenation](https://pandas.pydata.org/pandas-docs/stable/user_guide/merging.html#concatenating-objects), which allows us to connect dataframes together row-wise. 

<img src="assets/imgs/concat.png" width="40%"/>

Let's demonstrate this with an example. We'll create 3 small dataframes and use Pandas' [concat()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html#pandas.concat) function to join them together.

In [34]:
df1 = pd.DataFrame({'A': [1,10,15], 'B': [0.1,0.7,0.9], 'year':[2018, 2018, 2018]})
df2 = pd.DataFrame({'A': [4,18,29], 'B': [0.03,0.8,0.11], 'year':[2017, 2017, 2017]})
df3 = pd.DataFrame({'A': [17,30], 'B': [0.8,0.11], 'year':[2019, 2019]})

result = pd.concat([df1, df2, df3])
result

A     B  year
0   1  0.10  2018
1  10  0.70  2018
2  15  0.90  2018
0   4  0.03  2017
1  18  0.80  2017
2  29  0.11  2017
0  17  0.80  2019
1  30  0.11  2019

The concatenanted dataframe has as many rows as the smaller dataframes combined.

Let's go back to our Starbucks menu. We have been focusing on the beverages, but there is also a separate dataset with food items. Can we combine the food and drink menus into a single dataframe? Let's see if it's possible. 

In [35]:
starbucks_food = pd.read_csv("data/starbucks_food_menu.csv")

In [36]:
starbucks_food.head()

item  calories  fat_in_g  carbs_in_g  fiber_in_g  protein_in_g
0      Chonga Bagel       300       5.0          50           3            12
1      8-Grain Roll       380       6.0          70           7            10
2  Almond Croissant       410      22.0          45           3            10
3     Apple Fritter       460      23.0          56           2             7
4  Banana Nut Bread       420      22.0          52           2             6

We can see that there's some overlap in column names between the Starbucks food and beverage dataframes. The only column that is different is `item` in `starbucks_food`. We'll have to standardize this column name in `starbucks_drink` before proceeding to concatenation. First, we'll create a new dataframe called `sb_drink_subset` which will represent a subset of columns from our original `starbucks_drink` dataframe.

In [37]:
sb_drink_subset = starbucks_drink[['beverage','drink_size','calories','fat_in_g','carbs_in_g','fiber_in_g','protein_in_g']]

Let's create a new column that represents "item" in `sb_drink_subset`. We'll do this by combining "beverage" and "drink_size". We'll format it so that it looks like this:

```
beverage/drink_size
```


In [38]:
sb_drink_subset['item'] = sb_drink_subset['beverage'] + "/" + sb_drink_subset['drink_size']

We don't need our "beverage" and "drink_size" columns any more, so let's drop them from our dataframe. 

In [39]:
sb_drink_subset = sb_drink_subset.drop(columns=['beverage', 'drink_size'])

We can now concatenate `sb_drink_subset` with `starbucks_food` to create a dataframe containing all Starbucks menu items. 

In [40]:
starbucks_menu = pd.concat([sb_drink_subset, starbucks_food], sort=False)
print(f"Dataframe dimensions: {starbucks_menu.shape}")
starbucks_menu.head()

Dataframe dimensions: (353, 6)


calories  fat_in_g  carbs_in_g  fiber_in_g  protein_in_g  \
0         3       0.1           5           0           0.3   
1         4       0.1          10           0           0.5   
2         5       0.1          10           0           1.0   
3         5       0.1          10           0           1.0   
4         5       0.0           5           0           0.4   

                    item  
0    Brewed Coffee/short  
1     Brewed Coffee/tall  
2   Brewed Coffee/grande  
3    Brewed Coffee/venti  
4  Caffè Americano/short

Our newly created `starbucks_menu` dataframe consists of 353 rows, suggesting that there's 353 items (both food and beverage) on the full menu. 

We can write this dataframe to a csv file for later use. We'll use pandas' [to_csv](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_csv.html) to do this. We don't want to save the index as a column in our csv file, so we'll set `index` to be False.

In [41]:
starbucks_menu.to_csv("data/starbucks_menu.csv", index=False)